File: js/3.mjs
import { readFile } from "fs";

readFile("../inputs/3.txt", "utf8", (err, data) => {
    if (err) console.log(err);
    const input = data.split("\n");

    var zeroes = [0,0,0,0,0,0,0,0,0,0,0,0];
    var ones = [0,0,0,0,0,0,0,0,0,0,0,0];
    
    input.forEach((line, i) => {
        line.split("").forEach((char, ii) => {
            if (char === '0') {
                zeroes[ii]++;
            }
            if (char === '1') {
                ones[ii]++;
            }
        });
    });

    var final = [0,0,0,0,0,0,0,0,0,0,0,0]

    zeroes.forEach((num, i) => {
        if (num > ones[i]) {
            final[i] = '0';
        } else {
            final[i] = '1';
        }
    });

    console.log(final.join(''));

    var epsilon = [0,0,0,0,0,0,0,0,0,0,0,0];

    final.forEach((num, i) => {
        if (num === '0') {
            epsilon[i] = '1';
        } else {
            epsilon[i] = '0';
        }
    });

    console.log(epsilon.join(''));

    epsilon = parseInt(epsilon.join(''), 2);
    var gamma = parseInt(final.join(''), 2);

    console.log(epsilon*gamma);
    // finished at 23:07
})

File: go/3.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	var lines []string

	file, err := os.Open("../inputs/3.txt")
	if err != nil {
		panic(err)
	}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	file.Close()

	var zeroes, ones []int

	zeroes = make([]int, 12)
	ones = make([]int, 12)

	for _, line := range lines {
		for i, char := range line {
			if char == '0' {
				zeroes[i]++
			} else {
				if char == '1' {
					ones[i]++
				}
			}
		}
	}

	var final []string
	final = make([]string, 12)

	for i, z := range zeroes {
		if z > ones[i] {
			final[i] = "0"
		} else {
			final[i] = "1"
		}
	}

	rEpsilon := make([]string, 12)

	for i, v := range final {
		if v == "1" {
			rEpsilon[i] = "0"
		} else {
			rEpsilon[i] = "1"
		}
	}

	fmt.Println(strings.Join(rEpsilon, ""))
	fmt.Println(strings.Join(final, ""))

	epsilonInt, err := strconv.ParseInt(strings.Join(rEpsilon, ""), 2, 16)
	if err != nil {
		panic(err)
	}
	gammaInt, err := strconv.ParseInt(strings.Join(final, ""), 2, 16)
	if err != nil {
		panic(err)
	}

	fmt.Println(epsilonInt * gammaInt)

	// finished at 23:19
}
